File: policy-engine/pkg/fixtures/fixtures.go
// Package fixtures loads PBAC YAML config directories into shapes the
// engine can consume.
//
// The directory layout matches pbac/ at the repo root:
//
//	<config>/consumers/*.yml   top-level key: consumer:
//	<config>/purposes/*.yml    top-level key: purpose:
//	<config>/datasets/*.yml    fideslang Dataset YAML, top-level key: dataset:
//	<config>/policies/*.yml    top-level key: policy:
package fixtures

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/ethyca/fides/policy-engine/pkg/pbac"
	"gopkg.in/yaml.v3"
)

// Consumer is a data consumer loaded from YAML. Members is the list of
// identities (typically email addresses) that resolve to this consumer.
type Consumer struct {
	Name        string   `yaml:"name" json:"name"`
	Description string   `yaml:"description,omitempty" json:"description,omitempty"`
	Members     []string `yaml:"members" json:"members"`
	Purposes    []string `yaml:"purposes" json:"purposes"`
}

type consumerFile struct {
	Consumer []Consumer `yaml:"consumer"`
}

// Purpose is a declared purpose loaded from YAML.
type Purpose struct {
	FidesKey       string   `yaml:"fides_key" json:"fides_key"`
	Name           string   `yaml:"name" json:"name"`
	DataUse        string   `yaml:"data_use" json:"data_use"`
	DataSubject    string   `yaml:"data_subject,omitempty" json:"data_subject,omitempty"`
	DataCategories []string `yaml:"data_categories,omitempty" json:"data_categories,omitempty"`
	Description    string   `yaml:"description,omitempty" json:"description,omitempty"`
}

type purposeFile struct {
	Purpose []Purpose `yaml:"purpose"`
}

// dataset / collection / field carry data_purposes at each level. The
// engine reads dataset-level purposes directly and reads collection
// purposes via DatasetPurposes.CollectionPurposes. Field purposes are
// unioned into their owning collection's CollectionPurposes since the
// CLI extracts tables (collections) from SQL, not individual columns.
type dataset struct {
	FidesKey     string       `yaml:"fides_key"`
	Name         string       `yaml:"name,omitempty"`
	DataPurposes []string     `yaml:"data_purposes,omitempty"`
	Collections  []collection `yaml:"collections,omitempty"`
}

type collection struct {
	Name         string   `yaml:"name"`
	DataPurposes []string `yaml:"data_purposes,omitempty"`
	Fields       []field  `yaml:"fields,omitempty"`
}

type field struct {
	Name         string   `yaml:"name"`
	DataPurposes []string `yaml:"data_purposes,omitempty"`
}

type datasetFile struct {
	Dataset []dataset `yaml:"dataset"`
}

// Datasets bundles everything LoadDatasets returns: the per-dataset
// purpose map (engine input) and the table-name -> dataset_key index
// (table resolution).
type Datasets struct {
	// Purposes is keyed by dataset fides_key and fed to the engine.
	Purposes map[string]pbac.DatasetPurposes `json:"purposes"`
	// Tables maps a lowercase table name to its owning dataset's
	// fides_key. Assumes table names are globally unique across
	// datasets; on collision the last one loaded wins.
	Tables map[string]string `json:"tables"`
}

// LoadConsumers walks dir for *.yml files and returns a map from member
// identity to Consumer. A consumer with N members appears N times in the
// map, once per member, all pointing to the same Consumer value.
//
// When two consumers list the same member, the last one loaded wins.
// Returns an empty map (not an error) if dir doesn't exist.
func LoadConsumers(dir string) (map[string]Consumer, error) {
	out := map[string]Consumer{}
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		return out, nil
	}
	matches, err := filepath.Glob(filepath.Join(dir, "*.yml"))
	if err != nil {
		return nil, err
	}
	for _, path := range matches {
		var f consumerFile
		if err := readYAML(path, &f); err != nil {
			return nil, fmt.Errorf("consumers: %s: %w", path, err)
		}
		for _, c := range f.Consumer {
			for _, member := range c.Members {
				if member != "" {
					out[member] = c
				}
			}
		}
	}
	return out, nil
}

// LoadPurposes walks dir for *.yml files and returns a map keyed by FidesKey.
func LoadPurposes(dir string) (map[string]Purpose, error) {
	out := map[string]Purpose{}
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		return out, nil
	}
	matches, err := filepath.Glob(filepath.Join(dir, "*.yml"))
	if err != nil {
		return nil, err
	}
	for _, path := range matches {
		var f purposeFile
		if err := readYAML(path, &f); err != nil {
			return nil, fmt.Errorf("purposes: %s: %w", path, err)
		}
		for _, p := range f.Purpose {
			if p.FidesKey != "" {
				out[p.FidesKey] = p
			}
		}
	}
	return out, nil
}

// LoadDatasets walks dir for fideslang Dataset YAML files and returns
// both the per-dataset purpose map (for the engine) and a table ->
// dataset_key index built from collections[].name.
//
// Table resolution assumes collection names are globally unique across
// datasets, so SELECT ... FROM warehouse.orders and
// SELECT ... FROM archive.orders both resolve to whichever dataset
// declares a collection named "orders".
//
// Purposes are unioned across the three taxonomy levels:
//
//   - dataset.data_purposes                 → DatasetPurposes.PurposeKeys
//   - collection.data_purposes              ┐
//   - union(field.data_purposes for fields) ┴ → CollectionPurposes[name]
//
// The engine's EffectivePurposes(collection) then unions PurposeKeys
// with CollectionPurposes[collection], so the final effective set for
// a query on <dataset>.<collection> is the union of all three levels.
func LoadDatasets(dir string) (Datasets, error) {
	result := Datasets{
		Purposes: map[string]pbac.DatasetPurposes{},
		Tables:   map[string]string{},
	}
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		return result, nil
	}
	matches, err := filepath.Glob(filepath.Join(dir, "*.yml"))
	if err != nil {
		return result, err
	}
	for _, path := range matches {
		var f datasetFile
		if err := readYAML(path, &f); err != nil {
			return result, fmt.Errorf("datasets: %s: %w", path, err)
		}
		for _, ds := range f.Dataset {
			if ds.FidesKey == "" {
				continue
			}
			dp := pbac.DatasetPurposes{
				DatasetKey:         ds.FidesKey,
				PurposeKeys:        ds.DataPurposes,
				CollectionPurposes: map[string][]string{},
			}
			for _, c := range ds.Collections {
				if c.Name == "" {
					continue
				}
				name := strings.ToLower(c.Name)
				effective := collectionEffectivePurposes(c)
				if len(effective) > 0 {
					dp.CollectionPurposes[name] = effective
				}
				result.Tables[name] = ds.FidesKey
			}
			result.Purposes[ds.FidesKey] = dp
		}
	}
	return result, nil
}

// collectionEffectivePurposes unions a collection's own data_purposes
// with every one of its fields' data_purposes. Result is deduplicated
// and sorted for deterministic output.
func collectionEffectivePurposes(c collection) []string {
	set := map[string]bool{}
	for _, p := range c.DataPurposes {
		if p != "" {
			set[p] = true
		}
	}
	for _, fd := range c.Fields {
		for _, p := range fd.DataPurposes {
			if p != "" {
				set[p] = true
			}
		}
	}
	if len(set) == 0 {
		return nil
	}
	out := make([]string, 0, len(set))
	for p := range set {
		out = append(out, p)
	}
	sort.Strings(out)
	return out
}

type policyFile struct {
	Policy []pbac.AccessPolicy `yaml:"policy"`
}

// LoadPolicies walks dir for *.yml files and returns the enabled access
// policies. Disabled policies (enabled: false) are filtered out so the
// CLI doesn't have to care about the enabled flag later.
//
// Returns an empty slice (not an error) if dir doesn't exist.
func LoadPolicies(dir string) ([]pbac.AccessPolicy, error) {
	out := []pbac.AccessPolicy{}
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		return out, nil
	}
	matches, err := filepath.Glob(filepath.Join(dir, "*.yml"))
	if err != nil {
		return nil, err
	}
	for _, path := range matches {
		var f policyFile
		if err := readYAML(path, &f); err != nil {
			return nil, fmt.Errorf("policies: %s: %w", path, err)
		}
		for _, p := range f.Policy {
			if p.Enabled != nil && !*p.Enabled {
				continue
			}
			out = append(out, p)
		}
	}
	return out, nil
}

func readYAML(path string, into interface{}) error {
	b, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	return yaml.Unmarshal(b, into)
}

File: src/fides/service/pbac/fixtures.py
"""YAML fixture loaders for CLI-driven PBAC evaluation.

Mirrors policy-engine/pkg/fixtures/fixtures.go. The directory layout is
the one documented in pbac/README.md at the repo root:

    <config>/consumers/*.yml   top-level key: consumer:
    <config>/purposes/*.yml    top-level key: purpose:
    <config>/datasets/*.yml    fideslang Dataset YAML, top-level: dataset:
    <config>/policies/*.yml    top-level key: policy:

The loaders produce the same shapes as the Go loaders so the Python and
Go pipelines stay interchangeable per-statement. They are deliberately
dumb — no validation beyond "file parses as a YAML mapping" — because
the pipeline treats unknown identities, missing datasets, and empty
policy sets as first-class outcomes (gaps).
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml

from fides.service.pbac.policies.evaluate import ParsedPolicy, parsed_policy_from_dict
from fides.service.pbac.types import DatasetPurposes

# ── Dataclasses mirroring the Go structs ─────────────────────────────


@dataclass(frozen=True)
class Consumer:
    """Matches fixtures.Consumer on the Go side."""

    name: str
    description: str = ""
    members: tuple[str, ...] = ()
    purposes: tuple[str, ...] = ()


@dataclass(frozen=True)
class Purpose:
    """Matches fixtures.Purpose on the Go side.

    data_use is the single required taxonomy binding; the CLI uses it
    to map dataset purpose keys to data_use strings when evaluating
    access policies (see Pipeline.evaluate).
    """

    fides_key: str
    name: str
    data_use: str
    data_subject: str = ""
    data_categories: tuple[str, ...] = ()
    description: str = ""


@dataclass(frozen=True)
class Datasets:
    """Matches fixtures.Datasets on the Go side.

    purposes is keyed by dataset fides_key and fed to the engine.
    tables maps a lowercase table (collection) name to its owning
    dataset's fides_key.
    """

    purposes: dict[str, DatasetPurposes] = field(default_factory=dict)
    tables: dict[str, str] = field(default_factory=dict)


@dataclass(frozen=True)
class Fixtures:
    """Everything the pipeline consumes, pre-loaded.

    Built by load_fixtures(). Kept frozen so a single Fixtures instance
    can be shared across concurrent pipeline calls.
    """

    consumers: dict[str, Consumer] = field(default_factory=dict)
    purposes: dict[str, Purpose] = field(default_factory=dict)
    datasets: Datasets = field(default_factory=Datasets)
    policies: tuple[ParsedPolicy, ...] = ()


# ── Top-level loader ─────────────────────────────────────────────────


def load_fixtures(config_dir: str | os.PathLike[str]) -> Fixtures:
    """Load consumers/, purposes/, datasets/, policies/ from config_dir.

    Missing subdirectories are treated as empty (not an error); the
    pipeline's gap outputs cover the cases where that matters.
    """
    root = Path(config_dir)
    return Fixtures(
        consumers=load_consumers(root / "consumers"),
        purposes=load_purposes(root / "purposes"),
        datasets=load_datasets(root / "datasets"),
        policies=load_policies(root / "policies"),
    )


# ── Per-kind loaders ─────────────────────────────────────────────────


def load_consumers(directory: str | os.PathLike[str]) -> dict[str, Consumer]:
    """Return a map from member identity to Consumer.

    A consumer with N members appears N times in the map, once per
    member. Empty strings in `members` are skipped. When two consumers
    claim the same member, the last file loaded wins (fixture authors
    are expected to keep membership unambiguous).
    """
    directory = Path(directory)
    result: dict[str, Consumer] = {}
    if not directory.is_dir():
        return result
    for path in sorted(directory.glob("*.yml")):
        data = _read_yaml_mapping(path)
        for entry in data.get("consumer", []) or []:
            consumer = Consumer(
                name=entry.get("name", ""),
                description=entry.get("description", ""),
                members=tuple(entry.get("members", []) or []),
                purposes=tuple(entry.get("purposes", []) or []),
            )
            for member in consumer.members:
                if member:
                    result[member] = consumer
    return result


def load_purposes(directory: str | os.PathLike[str]) -> dict[str, Purpose]:
    """Return a map from fides_key to Purpose."""
    directory = Path(directory)
    result: dict[str, Purpose] = {}
    if not directory.is_dir():
        return result
    for path in sorted(directory.glob("*.yml")):
        data = _read_yaml_mapping(path)
        for entry in data.get("purpose", []) or []:
            key = entry.get("fides_key", "")
            if not key:
                continue
            result[key] = Purpose(
                fides_key=key,
                name=entry.get("name", ""),
                data_use=entry.get("data_use", ""),
                data_subject=entry.get("data_subject", ""),
                data_categories=tuple(entry.get("data_categories", []) or []),
                description=entry.get("description", ""),
            )
    return result


def load_datasets(directory: str | os.PathLike[str]) -> Datasets:
    """Return per-dataset purposes plus the table -> dataset_key index.

    Collection names are lowercased in the index so `orders` and
    `ORDERS` resolve to the same dataset. When multiple datasets declare
    a collection with the same name, the last file loaded wins.

    Purpose inheritance mirrors the Go loader:

        dataset.data_purposes
        -> DatasetPurposes.purpose_keys

        collection.data_purposes
        | union(field.data_purposes for each field in the collection)
        -> DatasetPurposes.collection_purposes[collection_name]

    The engine later unions these in effective_purposes(collection).
    """
    directory = Path(directory)
    result = Datasets(purposes={}, tables={})
    if not directory.is_dir():
        return result

    for path in sorted(directory.glob("*.yml")):
        data = _read_yaml_mapping(path)
        for entry in data.get("dataset", []) or []:
            fides_key = entry.get("fides_key", "")
            if not fides_key:
                continue
            collection_purposes: dict[str, frozenset[str]] = {}
            for collection in entry.get("collections", []) or []:
                name = (collection.get("name") or "").lower()
                if not name:
                    continue
                effective = _collection_effective_purposes(collection)
                if effective:
                    collection_purposes[name] = frozenset(effective)
                result.tables[name] = fides_key

            result.purposes[fides_key] = DatasetPurposes(
                dataset_key=fides_key,
                purpose_keys=frozenset(entry.get("data_purposes", []) or []),
                collection_purposes=collection_purposes,
            )

    return result


def load_policies(directory: str | os.PathLike[str]) -> tuple[ParsedPolicy, ...]:
    """Return the enabled policies from YAML files in load-sorted order.

    Each file's top-level `policy:` entries are merged into one list.
    Policies with `enabled: false` are filtered out — the pipeline
    never has to care about the enabled flag after load.
    """
    directory = Path(directory)
    result: list[ParsedPolicy] = []
    if not directory.is_dir():
        return tuple(result)
    for path in sorted(directory.glob("*.yml")):
        data = _read_yaml_mapping(path)
        for entry in data.get("policy", []) or []:
            if entry.get("enabled") is False:
                continue
            result.append(parsed_policy_from_dict(entry))
    return tuple(result)


# ── Private helpers ──────────────────────────────────────────────────


def _read_yaml_mapping(path: Path) -> dict[str, Any]:
    """Parse a YAML file that must contain a top-level mapping.

    Returns an empty dict for empty files or non-mapping documents —
    consistent with the Go loader's behavior of ignoring malformed
    files rather than aborting the whole load.
    """
    with path.open("r", encoding="utf-8") as f:
        data = yaml.safe_load(f)
    return data if isinstance(data, dict) else {}


def _collection_effective_purposes(collection: dict[str, Any]) -> list[str]:
    """Union a collection's own data_purposes with every field's.

    Result is deduplicated; order doesn't matter because the engine
    treats it as a set. Matches policy-engine/pkg/fixtures/fixtures.go:
    collectionEffectivePurposes().
    """
    purposes: set[str] = set(collection.get("data_purposes", []) or [])
    for fld in collection.get("fields", []) or []:
        for p in fld.get("data_purposes", []) or []:
            if p:
                purposes.add(p)
    purposes.discard("")
    return sorted(purposes)
